### ISSUE TITLE
Restart needed to display new band

### ISSUE THREAD

**Issue by fruitofgeorge** [berkon/wireless-microphone-analyzer#42]
Hi. Thanks for making this useful piece of software. 
I'm using the latest release on mac OS 12.6.9 and have noticed weird behavior. When I select a new band I can see on the rf explorer's display that it takes the new frequency range but the software displays nothing and reports the range as NaN - NaN Hz. If I quit and reopen the application then it adopts that new band the rfe is set to and displays correctly. This happens any time a new band is selected.
I'd also like to make a suggestion for a "manual" band which allows the user to input a start and stop frequency. This could be useful if you're using multiple bands or different brands mixed and want to see an overview of everything.

**Comment by berkon** [comment]
I'm very happy to hear that the tool is helpful for you. The thing with MacOS is, that I only have a very old (MacOS Sierra) machine which I'm using for DJing. It works very well and as I'm using it for DJing only, I don't want to upgrade to a newer MacOS. You know ... never change a running system. ;-) In general I wouldn't expect that the tool should behave differently on newer MacOS versions because the Electron/Node version stays the same for all systems, but of course I can't tell for sure. Nevertheless we have a few options now. First I would recommend to reset all settings (in "Tools" menu). Then if the issue is still there, please do the following: open the developer tools (in "Help" menu) and switch to the "Console" tab. Then please reproduce the issue and copy paste me the log output. BTW the next release will also contain a menu entry to collect the logs automatically.

Thanks for your suggestion regarding the frequency range. I'll put it on my list.

**Comment by fruitofgeorge** [comment]
I reset all settings and the issue persisted.
Here's the log output:

```
/Applications/Wireless Microphone Analyzer.app/Contents/Resources/app.asar/renderer.js:132 MX Linux workaround is disabled
/Applications/Wireless Microphone Analyzer.app/Contents/Resources/app.asar/renderer.js:140 Dark mode is disabled
/Applications/Wireless Microphone Analyzer.app/Contents/Resources/app.asar/renderer.js:371 Scan device is 'RF_EXPLORER'
/Applications/Wireless Microphone Analyzer.app/Contents/Resources/app.asar/node_modules/chart.js/dist/Chart.js:2432 bar chart: "scales.[x/y]Axes.barPercentage" is deprecated. Please use "dataset.barPercentage" instead
_deprecated @ /Applications/Wireless Microphone Analyzer.app/Contents/Resources/app.asar/node_modules/chart.js/dist/Chart.js:2432
/Applications/Wireless Microphone Analyzer.app/Contents/Resources/app.asar/renderer.js:1087 The following ports were found:
/Applications/Wireless Microphone Analyzer.app/Contents/Resources/app.asar/renderer.js:1090 [0] undefined (VendorID: undefined, ProductID: undefined)
/Applications/Wireless Microphone Analyzer.app/Contents/Resources/app.asar/renderer.js:1090 [1] undefined (VendorID: 10c4, ProductID: ea60)
/Applications/Wireless Microphone Analyzer.app/Contents/Resources/app.asar/renderer.js:1090 [2] undefined (VendorID: 10c4, ProductID: ea60)
/Applications/Wireless Microphone Analyzer.app/Contents/Resources/app.asar/renderer.js:1137 Connecting to dedicated port '/dev/tty.SLAB_USBtoUART'
/Applications/Wireless Microphone Analyzer.app/Contents/Resources/app.asar/renderer.js:646 =========================================================
/Applications/Wireless Microphone Analyzer.app/Contents/Resources/app.asar/renderer.js:647 Trying port /dev/tty.SLAB_USBtoUART with baud rate 500000 ...
/Applications/Wireless Microphone Analyzer.app/Contents/Resources/app.asar/renderer.js:674 Successfully connected to /dev/tty.SLAB_USBtoUART!
/Applications/Wireless Microphone Analyzer.app/Contents/Resources/app.asar/scan_devices/rf_explorer.js:73 Setting handler for RF Explorer data receiption ... 
/Applications/Wireless Microphone Analyzer.app/Contents/Resources/app.asar/renderer.js:860 Started response check timer 1
/Applications/Wireless Microphone Analyzer.app/Contents/Resources/app.asar/scan_devices/rf_explorer.js:36 Probing for 'RF Explorer' hardware (with cmd: '#C0' ) ...
/Applications/Wireless Microphone Analyzer.app/Contents/Resources/app.asar/scan_devices/rf_explorer.js:84 Received device name string 'uxÿþÿþRF Explorer 03.42 01-Nov-23 05.01.08'
/Applications/Wireless Microphone Analyzer.app/Contents/Resources/app.asar/renderer.js:711 Stoping response check timer 1 ...
/Applications/Wireless Microphone Analyzer.app/Contents/Resources/app.asar/renderer.js:714 Successfully detected 'RF Explorer' hardware!
/Applications/Wireless Microphone Analyzer.app/Contents/Resources/app.asar/scan_devices/rf_explorer.js:132 Received model, setup and firmware version:  ID: #C2-M: DATA LENGTH: 13 DATA: 010,012,03.42
/Applications/Wireless Microphone Analyzer.app/Contents/Resources/app.asar/scan_devices/rf_explorer.js:100 Received config data:  ID: #C2-F: DATA LENGTH: 75 DATA: 0470125,0646396,-020,-110,0112,0,000,0000050,0960000,0959950,00600,0000,000
/Applications/Wireless Microphone Analyzer.app/Contents/Resources/app.asar/scan_devices/rf_explorer.js:54 Setting frequency configuration to start: 0470000 stop: 0616000 ...
/Applications/Wireless Microphone Analyzer.app/Contents/Resources/app.asar/scan_devices/rf_explorer.js:100 Received config data:  ID: #C2-F: DATA LENGTH: 146 DATA: ª¢¬¦¢£¥©¨ ££§  §¢¤© ¡¥¥¦ÿþÿþ#C2-F:0470000,1315314,-020,-110,0112,0,000,0000050,0960000,0959950,00600,0000,000
```

You can see that last line is right after I select a different band. Clearly has some junk data preceeding the real data.

**Comment by berkon** [comment]
Ok, so the junk data at the end is most likely not really junk, but instead it is already scan data, which arrives and is not expected there to come. There are three potential reasons for the issue.

- I saw that you have the RF-Explorer model WSUB1G_PLUS. I have the (much) older one WSUB1G. As from my understanding the firmware is the same for all models, I wouldn't expect that the newer HW itself causes this behavior.

- I have 5 year old firmware on my device. So maybe the newer firmware differs in behavior. The problem is that I only have one hardware. So if I update it to a newer version, I cannot make sure that it still works for users with older firmware versions.

- Finally of course there could still be a software bug ;-)

Let me think about it ... ;-)

One last question: Is the issue reproducible in 100% of the cases?

**Comment by fruitofgeorge** [comment]
Yes, I do have the plus model. And I believe the firmware is different between the normal and plus models as well. At least the versioning is different. Current for standard model is 1.34 while current plus model is 3.42. Looking over the API, they did add data to the #C2-F response in v1.09 and even more in v1.12. However, the additional data is added to the tail of the response, not at the head as the log shows. Also the log is showing the full response is interpreted correctly according to the latest firmware following the garbled data.

Yes, it is reproducible 100% of the time I select a new band from the menu. Although the data length reported in the logs does vary seemingly random. A normal response seems like it should be 75 bytes but I've seen as little as 83 bytes and as much as 193 bytes including the additional garbled data.

**Comment by berkon** [comment]
Ok, then indeed it looks like an issue with the different firmware. I tried a few times yesterday, but I was not able to see the issue, not even a single time. But I guess I have good news! Yesterday I got an email from the manufacturer company. They were reading our posts in this issue and they offered to send a test device to me, so that I can fix the bug. I'll let you know as soon as I have any news on this.

**Comment by fruitofgeorge** [comment]
That is great news! What a kind gesture from them. (Unfortunately their RF Explorer Touch VM image on a mac is abysmal.)

**Comment by berkon** [comment]
@fruitofgeorge, I haven't received the new hardware yet, but I have an idea which might help. I've created a test build, so you can try if it works. What I'm doing now, is to tell the device explicitly to hold the incoming data stream before sending a new command. This sounds obvious, but interestingly it was not necessary for the old device. And I cannot promise that it solve the isssue, but I think its worth a try:

https://drive.google.com/drive/folders/15BS0QXSAVlEdTf7O35w1g6d8eFS2zNyl?usp=sharing

**Comment by fruitofgeorge** [comment]
I appreciate you continuing to work on this but unfortunately that didn't solve the issue. The behavior is exactly the same and the response in the logs looks the same.
I do like the new manual entry band. Seems to work nicely aside from the fact that it is perceived as a new band and requires a reset the same as any band.

**Comment by berkon** [comment]
Oh, that's a pity. Ok, then I'll have to wait for the device to arrive.

I also added hotkeys (number keys 1 - 9) which can be used to quickly switch between bands. Saving a hotkey can be done either via the menu or even simpler by pressing <SHIFT><hotkey>. I think that could be quite useful too. But in your case of course it will only unfold full potential once the issue is fixed.

**Comment by berkon** [comment]
Good news! I've just received the device and I can confirm that I can reproduce the issue on Windows too. I'll let you know as soon as I found the problem.

**Comment by berkon** [comment]
I've just released version V2.1.0. The issue should be fixed now.
